Extensions/TokenAcquisition.cs

```diff
@@ -0,0 +1,188 @@
+                var request = context.HttpContext.Request;
+                var currentUri = UriHelper.BuildAbsolute(request.Scheme, request.Host, request.PathBase, request.Path);
+                var credential = new ClientCredential(azureAdOptions.ClientSecret);
+                Application = new ConfidentialClientApplication(azureAdOptions.ClientId, currentUri, credential, AuthPropertiesTokenCacheHelper.ForCodeRedemption(context.Properties), null);
+
+                var result = await Application.AcquireTokenByAuthorizationCodeAsync(context.ProtocolMessage.Code, scopes);
+                context.HandleCodeRedemption(result.AccessToken, result.IdToken);
+            }
+            catch (MsalException ex)
+            {
+                string message = ex.Message;
+                throw;
+            }
+
+        }
+
+
+        /// <summary>
+        /// Gets an access token for a downstream API on behalf of the user which evidence are provided by the
+        /// <paramref name="user"/> parameter
+        /// </summary>
+        /// <param name="context">HttpContext (for instance of the controller)</param>
+        /// <param name="user">Account described by its claims</param>
+        /// <param name="scopes">Scopes to request for the downstream API to call</param>
+        /// <returns>An access token to call the downstream API characterized by its scopes, on behalf of the user</returns>
+        public async Task<string> GetAccessTokenOnBehalfOfUser(HttpContext context, ClaimsPrincipal user, string[] scopes)
+        {
+            // Use MSAL to get the right token to call the API
+            var credential = new ClientCredential(azureAdOptions.ClientSecret);
+            var request = context.Request;
+            var currentUri = UriHelper.BuildAbsolute(request.Scheme, request.Host, request.PathBase, request.Path);
+            Application = new ConfidentialClientApplication(azureAdOptions.ClientId, currentUri, new ClientCredential(azureAdOptions.ClientSecret), 
+                AuthPropertiesTokenCacheHelper.ForApiCalls(context, CookieAuthenticationDefaults.AuthenticationScheme), null);
+
+            string userObjectId = user.FindFirst("http://schemas.microsoft.com/identity/claims/objectidentifier").Value;
+            string tenantId = user.FindFirst("http://schemas.microsoft.com/identity/claims/tenantid").Value;
+            if (string.IsNullOrWhiteSpace(userObjectId))
+            {
+                // TODO: find a better typed exception
+                throw new Exception("Missing claim 'http://schemas.microsoft.com/identity/claims/objectidentifier'");
+            }
+
+            if (string.IsNullOrWhiteSpace(tenantId))
+            {
+                throw new Exception("Missing claim 'http://schemas.microsoft.com/identity/claims/tenantid'");
+            }
+            string accountId = userObjectId + "." + tenantId;
```

Comment:
line break after if clause.
no need to use '{' for single line clause.

Reply:
The Cambridge team insisted that I do … so ...

Reply:
I changed this not to use "{". Let me know if this needs to be changed back.